File: kotlin/src/main/kotlin/io/github/jkjamies/automationagent/agent/setup/Prompts.kt
/*
 * Package setup holds shared utilities for building agents: the LLM provider switch and
 * adapters, the prompt loader, and ADK helpers. It is the only package permitted to import
 * provider SDKs (Ollama, Gemini) — enforced by the :konsist arch rules.
 */
package io.github.jkjamies.automationagent.agent.setup

/**
 * Loads markdown prompt files from the classpath. Each agent keeps its prompts under
 * `src/main/resources/prompts/<agent>/<name>.md`, so they stay reviewable markdown next to
 * the agent that uses them.
 */
class Prompts(
    private val basePath: String,
    private val loader: ClassLoader = Prompts::class.java.classLoader,
) {
    /**
     * Returns the trimmed contents of `<basePath>/<name>.md`. Throws
     * [IllegalArgumentException] if the resource is absent — call at agent construction time,
     * where a missing prompt is a programming error that should fail fast at startup.
     */
    fun get(name: String): String =
        getOrNull(name) ?: throw IllegalArgumentException("read prompt \"$name\": resource not found at $basePath/$name.md")

    /** Like [get] but returns null instead of throwing when the prompt is absent. */
    fun getOrNull(name: String): String? =
        loader.getResourceAsStream("$basePath/$name.md")
            ?.use { it.readBytes().toString(Charsets.UTF_8).trim() }

    companion object {
        /** Prompts rooted at `prompts/<agent>` (e.g. `forAgent("summary").get("summarize")`). */
        fun forAgent(agent: String): Prompts = Prompts("prompts/$agent")
    }
}

File: javascript/src/agent/setup/prompts.ts
/**
 * Markdown prompt loader.
 *
 * Each agent ships its own `prompts/` directory of markdown files and reads them
 * through {@link Prompts}, so prompts stay reviewable next to the agent that uses
 * them. Files are read from disk relative to the agent's directory, resolved from
 * `import.meta.url`.
 */
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

/** Loads `prompts/<name>.md` from a given anchor directory. */
export class Prompts {
  /**
   * @param anchorDir - the directory whose `prompts/` subdir holds the files.
   *   Agents pass `dirname(fileURLToPath(import.meta.url))`.
   */
  constructor(private readonly anchorDir: string) {}

  /**
   * Return the trimmed contents of `prompts/<name>.md`.
   *
   * @throws Error if the prompt file is missing or unreadable.
   */
  get(name: string): string {
    const path = join(this.anchorDir, 'prompts', `${name}.md`);
    try {
      return readFileSync(path, 'utf-8').trim();
    } catch (err) {
      throw new Error(`read prompt ${JSON.stringify(name)}: ${(err as Error).message}`);
    }
  }

  /**
   * Like {@link get}, but intended for agent-construction time where a missing
   * prompt is a programming error that should fail fast at startup.
   */
  mustGet(name: string): string {
    return this.get(name);
  }
}
